read.go

```diff
@@ -144,27 +145,13 @@ func (r *reader) readIndexData(toc *indexTOC) (*indexData, error) {
 		branchNames:    []map[uint]string{},
 	}
 
-	blob, err := d.readSectionBlob(toc.metaData)
-	if err != nil {
-		return nil, err
-	}
-
-	if err := json.Unmarshal(blob, &d.metaData); err != nil {
-		return nil, err
-	}
-
-	if d.metaData.IndexFormatVersion != IndexFormatVersion {
-		return nil, fmt.Errorf("file is v%d, want v%d", d.metaData.IndexFormatVersion, IndexFormatVersion)
-	}
-
-	blob, err = d.readSectionBlob(toc.repoMetaData)
-	if err != nil {
-		return nil, err
-	}
-
-	d.repoMetaData = make([]Repository, 1)
-	if err := json.Unmarshal(blob, &d.repoMetaData[0]); err != nil {
+	if repo, md, err := r.readMetadata(toc); md != nil && md.IndexFormatVersion != IndexFormatVersion {
+		return nil, fmt.Errorf("file is v%d, want v%d", md.IndexFormatVersion, IndexFormatVersion)
+	} else if err != nil {
 		return nil, err
+	} else {
+		d.metaData = *md
+		d.repoMetaData = []Repository{*repo}
```

Comment:
Should we flatten this?
```suggestion
	repo, md, err := r.readMetadata(toc); 
	if md != nil && md.IndexFormatVersion != IndexFormatVersion {
		return nil, fmt.Errorf("file is v%d, want v%d", md.IndexFormatVersion, IndexFormatVersion)
	} 
	if err != nil {
		return nil, err
	}
	d.metaData = *md
	d.repoMetaData = []Repository{*repo}
```

Reply:
Yeah this is how I would normally write it. But I thought it was nice avoiding `md` and `repo` getting into the function scope. But very minor, will flatten.